agent/engine/task_manager.go

```diff
@@ -481,6 +481,45 @@ func (mtask *managedTask) handleContainerChange(containerChange dockerContainerC
 	// Container has progressed its status if we reach here. Make sure to save it to database.
 	defer mtask.engine.saveContainerData(container)
 
+	// If container is transitioning to STOPPED, first check if we should short-circuit
+	// the stop workflow and restart the container.
+	if event.Status == apicontainerstatus.ContainerStopped && container.RestartPolicyEnabled() {
+		exitCode := event.DockerContainerMetadata.ExitCode
+		shouldRestart, reason := container.RestartTracker.ShouldRestart(exitCode, container.GetStartedAt(), container.GetDesiredStatus())
+		if shouldRestart {
+			defer container.RestartTracker.RecordRestart()
+			logger.Info("Restarting container", eventLogFields,
```

Comment:
Do we mean to not include this current restart attempt in the `restartCount` or `lastRestartAt`? It feels a bit odd to have a log that reads "Restarting container, restartCount=0"

Reply:
Yes that was intended, if you log before restarting then you havent restarted yet, and then there is another log line after the restart happens below.

I should maybe change it to increment before logging the "Restarted" log line below though 🤔

Reply:
Yeah the actual container hasn't been restarted yet, but we've begun another "restart attempt" is probably how I feel 
Maybe we can rename to `previousRestartCount`?

Reply:
since we will log after doing the restart no matter what, I will actually just delete this log line as it's a bit redundant anyways